cmd/pbstatuser/main.go

```diff
@@ -20,25 +26,179 @@ import (
-	// 4) Reads from GCP channel, performs availability check, has configurable throttling.
+	// Fetch authentication from Sources
+	authentication, err := sourcesClient.GetAuthentication(ctx, sourceId)
+	if err != nil {
+		if errors.Is(err, clients.NotFoundErr) {
+			// TODO: Add error handling
+			return
+		}
+		// TODO: Add error handling
+		return
+	}
+
+	switch authentication.ProviderType {
+	case models.ProviderTypeAWS:
+		chAws <- Pipeline{
+			Authentication: authentication,
+			SourceID:       sourceId,
+		}
+	case models.ProviderTypeAzure:
+		chAzure <- Pipeline{
+			Authentication: authentication,
+			SourceID:       sourceId,
+		}
+	case models.ProviderTypeGCP:
+		chGcp <- Pipeline{
+			Authentication: authentication,
+			SourceID:       sourceId,
+		}
+	case models.ProviderTypeNoop:
 
-	// Beware there is no database connection, no job queue, most of what api/worker processes have
-	// is not available because it should not be needed. The worker will simply use SDKs to check
-	// statuses and send out results via kafka.Send function to Sources.
-	// Also make sure that all goroutines are closed correctly on context cancel. A WaitGroup
-	// would be a really good here: https://gist.github.com/fracasula/b579d52daf15426e58aa133d0340ccb0
+	case models.ProviderTypeUnknown:
+		// TODO: Complete
+	}
+}
+
+func checkSourceAvailabilityAzure(ctx context.Context) {
+	defer processingWG.Done()
+
+	for p := range chAzure {
+		// TODO: check if source is avavliable
+		p.Status = StatusUnavailable
+		chSend <- p
+	}
+}
+
+func checkSourceAvailabilityAWS(ctx context.Context) {
+	defer processingWG.Done()
+
+	for p := range chAws {
+		_, err := clients.GetEC2Client(ctx, p.Authentication, "us-east-1")
+		if err != nil {
+			p.Status = StatusUnavailable
+			p.Err = err
+			chSend <- p
+		}
+		p.Status = StatusAvaliable
+		chSend <- p
+	}
+}
+
+func checkSourceAvailabilityGCP(ctx context.Context) {
+	defer processingWG.Done()
+
+	for p := range chGcp {
+		gcpClient, err := clients.GetGCPClient(ctx, p.Authentication)
+		if err != nil {
+			p.Status = StatusUnavailable
+			p.Err = err
+			chSend <- p
+		}
+		_, err = gcpClient.ListAllRegions(ctx)
+		if err != nil {
+			if err != nil {
+				p.Status = StatusUnavailable
+				p.Err = err
+				chSend <- p
+			}
+		}
+		p.Status = StatusAvaliable
+		chSend <- p
+	}
+}
+
+func sendResults(ctx context.Context) {
+	id := ctxval.Identity(ctx)
+	for p := range chSend {
+		p.ResourceType = "Application"
+		values, err := json.Marshal(p)
+		if err != nil {
+			// TODO: Add error handling
+			return
+		}
+		msg := &kafka.GenericMessage{
+			Topic: kafka.AvailabilityStatusRequestTopic,
+			Value: []byte(values),
+			Headers: map[string]string{
+				"Content-Type":                "application/json",
+				"x-rh-identity":               identity.GetIdentityHeader(ctx),
+				"x-rh-sources-org-id":         id.Identity.OrgID,
+				"x-rh-sources-account-number": id.Identity.AccountNumber,
+			},
+		}
```

Comment:
The headers creation needs to be moved into the `kafka` package. Now we need a "constructor", something like `NewGenericMessage` that will take context, topic and value and return a generic message with all the platform headers set correctly.

Then you want to use this in the sending code (in the api process) to send the headers as well. Right now they will be empty as nothing is really sent into our internal topic.

I have an extra ask, can you please use "X-RH-Sources-Account-Number" case? Headers (HTTP/Kafka) are not case sensitive, but it looks much nicer and it is the "defacto" standard convention for HTTP headers.

Reply:
Do you mean to add the headers here as well https://github.com/RHEnVision/provisioning-backend/blob/75bc8471cff0ffbb30908ad0bfec125fba3216e6/internal/kafka/avail_status_msg.go#L31? or to use the new function I created here internal/kafka/kafka.go

Reply:
Yes, headers need to be handled properly:

* sending it from the `services.AvailabilityStatus` (currently this is completely ignored and no headers are set)
* passing headers from/to GenericMessage to KafkaMessage
* sending it from the statuser send function

The type we use for ASM is currently this structure:

```
type AvailabilityStatusMessage struct {
	SourceID string `json:"source_id"`
}
```

It has no headers obviously, I suggest to refactor `GenericMessage(identity identity.XRHID)` function to get a context and construct headers from our own context (we should carry the whole identity there - see `ctxval.Identity()`).

The other direction is more tricky, one idea I have is to change the `NewAvailabilityStatusMessage` function to return three results: ASM, identity.XRHID and error, so you can then use the identity when constructing the result message back to sources (with the function above `GenericMessage(identity)`.